test/cljs/sieppari/core_execute_test.cljc

```diff
@@ -0,0 +1,194 @@
+;; idea is that the tests override the expected stage functions
+;; with test specific function. This ensures that no unexpected
+;; stage functions are called.
+;;
+
+; Make an interceptor with given name and set all stage functions
+; to report unexpected invocation. Tests should override expected
+; stages.
+
+(defn unexpected [name stage]
+  (fn [ctx]
+    (throw (ex-info "unexpected invocation"
+                    {:name name
+                     :stage stage
+                     :ctx ctx}))))
+
+(defn make-test-interceptor [name]
+  {:name name
+   :enter (unexpected name :enter)
+   :leave (unexpected name :leave)
+   :error (unexpected name :error)})
+
+; Test stack with three interceptors and a handler:
+
+(def test-chain [(make-test-interceptor :a)
+                 (make-test-interceptor :b)
+                 (make-test-interceptor :c)
+                 (unexpected :handler nil)])
+
+(def a-index 0)
+(def b-index 1)
+(def c-index 2)
+(def h-index 3)
+
+;; Helper: always throws an exception with specific marker
+;; in data part:
+
+(def error (ex-info "oh no" {::error-marker true}))
+
+(defn always-throw [ctx]
+  (throw error))
+
+;; Helper: return error handler function that ensures
+;; that `ctx` contains an exception caused by `always-throw`,
+;; clears the exception and sets response to given response:
+
+(defn handle-error [response]
+  (fn [ctx]
+    (assert (not (some? (:response ctx))))
+    (assert (-> ctx :error ex-data (= {::error-marker true})))
+    (-> ctx
+        (dissoc :error)
+        (assoc :response response))))
+
+;;
+;; Tests:
+;;
+
+(deftest happy-case-test
+  (async done
+    (let [res (atom nil)]
+      (-> test-chain
+          (assoc-in [a-index :enter] identity)
+          (assoc-in [b-index :enter] identity)
+          (assoc-in [c-index :enter] identity)
+          (assoc-in [h-index] inc)
+          (assoc-in [c-index :leave] identity)
+          (assoc-in [b-index :leave] identity)
+          (assoc-in [a-index :leave] identity)
+          (s/execute 41
+                     (partial reset! res)
+                     (partial reset! res)))
+      (is (= 42 @res) "enable all enter and leave stages, use `inc` as handler"))
+    (done)))
+
+#_(deftest enter-b-causes-exception-test
+    (fact ":b causes an exception"
+      (-> test-chain
+          (assoc-in [a-index :enter] identity)
+          (assoc-in [b-index :enter] always-throw)
+          (assoc-in [b-index :error] identity)
+          (assoc-in [a-index :error] identity)
+          (s/execute 41))
+      =throws=> error))
+
+#_(deftest enter-c-causes-exception-a-handles-test
+    (fact ":c enter causes an exception, :b sees error, :a handles"
+      (-> test-chain
+          (assoc-in [a-index :enter] identity)
+          (assoc-in [b-index :enter] identity)
+          (assoc-in [c-index :enter] always-throw)
+          (assoc-in [c-index :error] identity)
+          (assoc-in [b-index :error] identity)
```

Comment:
@Deraen @jarppe I have a Q on this test. In Cljs it throws a unexpected invocation on :b :leave.
I wonder why it does not throw in the Clj tests....what am I missing?

Reply:
Never mind, I think I am missing to reverse execute the :leave handlers